apps/backend/src/routes/socket/index.ts

```diff
@@ -0,0 +1,120 @@
+import { Server, Socket } from "socket.io";
+import { server } from "@/index";
+import {
+  SocketRequestType,
+  SocketRequestSchema,
+  SocketRequest,
+  SocketResponseType,
+  SocketResponse,
+  SocketErrorPayload,
+  MapRequestToResponse,
+  errorToString
+} from "@types";
+import { Controller } from "@/lib/controller";
+
+export const wsServer = new Server(server);
+
+const controller = new Controller();
+
+const clientsSockets: Record<string, Socket> = {};
+
+const handleError = (socket: Socket, recipient: string, error: string): void => {
+  const payload: SocketErrorPayload = {
+    error,
+  }
+
+  const resp: SocketResponse = {
+    type: SocketResponseType.ERROR,
+    recipientSigPubKey: recipient,
+    senderSigPubKey: "",
+    payload
+  };
+
+  socket.send(JSON.stringify(resp))
+  return;
+}
+
+wsServer.on('connection', (socket: Socket) => {
+  socket.on('open', () => {
+    console.log("Open socket client connection.");
+  });
+
+  socket.on('disconnect', () => {
+    console.log(`Disconnect socket client connection.`);
+  });
+
+  socket.on('message', async (message: string) => {
+    let sender: string | null = null;
+    let recipient: string | null = null;
+    try {
+      const req: SocketRequest = SocketRequestSchema.parse(JSON.parse(message));
+
+      if (!req.senderSigPubKey) {
+        throw new Error("Missing sender");
+      }
+
+      sender = req.senderSigPubKey;
+      recipient = req.recipientSigPubKey;
+
+      if (!req.authToken) {
+        return handleError(socket, sender, "Missing auth token.");
+      }
+
+      // Fetch user by auth token, ensure the request is from an authenticated user
+      const user = await controller.GetUserByAuthToken(req.authToken);
+      if (!user) {
+        return handleError(socket, sender, "Invalid auth token.");
+      } else if (user.signaturePublicKey !== sender) {
+        return handleError(socket, sender, "Incorrect user key");
+      }
+
+      switch (req.type) {
+        case SocketRequestType.CONNECT:
+          if (!sender) {
+            return handleError(socket, sender, "Missing sender.");
+          }
+
+          // Set record for future lookup
+          clientsSockets[sender] = socket;
+
+          // Return CONNECT_SUCCESS response
+          const connectResponse: SocketResponse = MapRequestToResponse(req);
+
+          if (clientsSockets[sender]) {
+            clientsSockets[sender].send(JSON.stringify(connectResponse));
+          } else {
+            throw new Error("Sender client socket missing.");
+          }
+          return;
+        case SocketRequestType.TAP_BACK:
```

Comment:
Again not urgent, but in the future having the sender send a tap back notif via websockets means the server can see this activity. a more private flow is that the server just notifies a client itself when it sees it has a new message (where sender privacy is provided via ephemeral ecdh)

Reply:
Unfortunately I don't understand the alternative you're suggesting. Is the idea that the server would poll for new messages, and when it gets a new message it would see if there's an active client socket, else skip?

Reply:
Nah, when the server receives new messages, it just sends a ws ping to all the recipients. but this is not necessary for now

Reply:
https://github.com/cursive-team/connections/pull/191/files#diff-e1f99bd817c2c6047065d9e66074644a68ab0e614021785848519a46c26cf730R99 -- isn't this only sending it to the recipient socket?